File: ios/opensleep/opensleep/Services/Sleep/AdaptiveNormalizationFilter.swift
import Foundation

/// Port of com.urbandroid.sleep.domain.AdaptiveNormalizationFilter from Sleep as Android.
enum AdaptiveNormalizationFilter {

    final class ANFResult {
        let input: [Float]
        let output: [Float]
        let median: Double
        let max: Double
        let exponent: Double

        init(input: [Float], output: [Float], median: Double, max: Double, exponent: Double) {
            self.input = input
            self.output = output
            self.median = median
            self.max = max
            self.exponent = exponent
        }

        func aggregateOutput(_ i: Int) -> [Float] {
            let length = output.count
            guard length > 0 else { return [] }
            let maxFunc = Moving.max(period: i)
            let iCeil = Int(ceil(Double(length) / Double(i)))
            var result = [Float](repeating: 0.0, count: iCeil)
            var lastApply: Float = 0.0

            for i2 in 0..<length {
                lastApply = maxFunc(output[i2])
                if i2 % i == i - 1 {
                    result[i2 / i] = lastApply
                }
            }
            result[iCeil - 1] = lastApply
            return result
        }

        func getHighActivityFlags(threshold: Float) -> IndexSet {
            var indexSet = IndexSet()
            for (idx, val) in output.enumerated() {
                if val > threshold {
                    indexSet.insert(idx)
                }
            }
            return indexSet
        }

        func getHighActivityFlagsFloat(threshold: Float) -> [Float] {
            output.map { $0 > threshold ? 1.0 : 0.0 }
        }
    }

    private static func getNoiseAmplitudeEstimate(_ arr: [Float], _ excluded: IndexSet) -> Float {
        var list: [Float] = []
        list.reserveCapacity(arr.count)
        for i in 0..<arr.count {
            if arr[i] >= 0.0 && !excluded.contains(i) {
                list.append(arr[i])
            }
        }
        guard list.count >= 3 else { return 0.0 }

        let percentile = Percentile()
        let e50 = percentile.evaluate(list, 50)
        if e50 > 0.0 { return e50 }
        let e70 = percentile.evaluate(list, 70)
        if e70 > 0.0 { return e70 }
        let e90 = percentile.evaluate(list, 90)
        if e90 > 0.0 { return e90 }
        let e95 = percentile.evaluate(list, 95)
        if e95 > 0.0 { return e95 }
        return 0.0
    }

    private static func getConservativeUpperBound(_ arr: [Float], _ excluded: IndexSet) -> Float {
        var copy = arr
        for idx in excluded {
            if idx < copy.count {
                copy[idx] = 1.0
            }
        }
        let q: Float = arr.count < 1000 ? 99.7 : (arr.count < 2500 ? 99.8 : 99.9)
        return Percentile().evaluate(copy, q)
    }

    static func normalizeAmplitudes(_ arr: [Float], excluded: IndexSet = IndexSet()) -> ANFResult {
        let length = arr.count
        guard length > 0 else {
            return ANFResult(input: arr, output: arr, median: 0.0, max: 0.0, exponent: 1.0)
        }

        var copy = arr
        let noiseEstimate = getNoiseAmplitudeEstimate(arr, excluded)
        if noiseEstimate == 0.0 {
            return ANFResult(input: arr, output: copy, median: 0.0, max: 0.0, exponent: 1.0)
        }

        for i in 0..<length {
            if copy[i] >= 0.0 {
                copy[i] /= noiseEstimate
            }
        }

        let upperBound = getConservativeUpperBound(copy, excluded)
        if upperBound <= 1.0 {
            return ANFResult(input: arr, output: copy, median: Double(noiseEstimate), max: 0.0, exponent: 1.0)
        }

        let dUpperBound = Double(upperBound)
        let exponent = 1.0 / log10(dUpperBound)
        for i in 0..<length {
            let v = copy[i]
            if v >= 0.0 {
                let clamped = min(upperBound, v)
                copy[i] = Float(pow(Double(clamped), exponent))
            }
        }

        return ANFResult(input: arr, output: copy, median: Double(noiseEstimate), max: dUpperBound, exponent: exponent)
    }
}

File: android/app/src/main/java/app/opensleep/domain/sleep/AdaptiveNormalizationFilter.kt
package app.opensleep.domain.sleep

import java.util.Arrays
import java.util.BitSet

/**
 * Port of com.urbandroid.sleep.domain.AdaptiveNormalizationFilter from Sleep as Android.
 */
object AdaptiveNormalizationFilter {

    class ANFResult(
        val input: FloatArray,
        val output: FloatArray,
        val median: Double,
        val max: Double,
        val exponent: Double
    ) {
        fun aggregateOutput(i: Int): FloatArray {
            val length = this.output.size
            if (length == 0) return FloatArray(0)
            val floatFunctionMax = Moving.max(i)
            val iCeil = Math.ceil(length.toDouble() / i.toDouble()).toInt()
            val fArr = FloatArray(iCeil)
            var fApply = 0.0f
            for (i2 in 0 until length) {
                fApply = floatFunctionMax.apply(this.output[i2])
                if (i2 % i == i - 1) {
                    fArr[i2 / i] = fApply
                }
            }
            fArr[iCeil - 1] = fApply
            return fArr
        }

        fun getHighActivityFlags(threshold: Float): BitSet {
            val bitSet = BitSet()
            for (idx in output.indices) {
                if (output[idx] > threshold) {
                    bitSet.set(idx)
                }
            }
            return bitSet
        }

        fun getHighActivityFlagsFloat(threshold: Float): FloatArray {
            val fArr = FloatArray(output.size)
            for (idx in output.indices) {
                if (output[idx] > threshold) {
                    fArr[idx] = 1.0f
                }
            }
            return fArr
        }
    }

    private fun getNoiseAmplitudeEstimate(fArr: FloatArray, bitSet: BitSet): Float {
        val list = ArrayList<Float>(fArr.size)
        for (i in fArr.indices) {
            if (fArr[i] >= 0.0f && !bitSet.get(i)) {
                list.add(fArr[i])
            }
        }
        if (list.size < 3) return 0.0f

        val percentile = Percentile()
        val e50 = percentile.evaluate(list, 50)
        if (e50 > 0.0f) return e50
        val e70 = percentile.evaluate(list, 70)
        if (e70 > 0.0f) return e70
        val e90 = percentile.evaluate(list, 90)
        if (e90 > 0.0f) return e90
        val e95 = percentile.evaluate(list, 95)
        if (e95 > 0.0f) return e95
        return 0.0f
    }

    private fun getConservativeUpperBound(fArr: FloatArray, bitSet: BitSet): Float {
        val length = fArr.size
        val copy = Arrays.copyOf(fArr, length)
        var i = 0
        while (true) {
            val nextSet = bitSet.nextSetBit(i)
            if (nextSet < 0) break
            if (nextSet < length) {
                copy[nextSet] = 1.0f
            }
            i = nextSet + 1
        }
        val q = if (length < 1000) 99.7f else if (length < 2500) 99.8f else 99.9f
        return Percentile().evaluate(copy, q)
    }

    fun normalizeAmplitudes(fArr: FloatArray, bitSet: BitSet = BitSet()): ANFResult {
        val length = fArr.size
        val copy = Arrays.copyOf(fArr, length)
        val noiseEstimate = getNoiseAmplitudeEstimate(fArr, bitSet)
        if (noiseEstimate == 0.0f) {
            return ANFResult(fArr, copy, 0.0, 0.0, 1.0)
        }

        for (i in 0 until length) {
            if (copy[i] >= 0.0f) {
                copy[i] /= noiseEstimate
            }
        }

        val upperBound = getConservativeUpperBound(copy, bitSet)
        if (upperBound <= 1.0f) {
            return ANFResult(fArr, copy, noiseEstimate.toDouble(), 0.0, 1.0)
        }

        val dUpperBound = upperBound.toDouble()
        val exponent = 1.0 / Math.log10(dUpperBound)
        for (i in 0 until length) {
            val v = copy[i]
            if (v >= 0.0f) {
                val clamped = Math.min(upperBound, v)
                copy[i] = Math.pow(clamped.toDouble(), exponent).toFloat()
            }
        }

        return ANFResult(fArr, copy, noiseEstimate.toDouble(), dUpperBound, exponent)
    }
}
